### ISSUE TITLE
Old Git commit used as input for job

### ISSUE THREAD

**Issue by robwhitby** [springernature/halfpipe#12]
https://springernature.slack.com/archives/C9LFM8B6X/p1559920305085500

From Vesa: hello, we noticed something a tad odd with one of our builds. Basically we had a failing build which we pushed a fix to. Instead of that commit being run Concourse seemingly resurrected a git commit from the end of last month and ran that instead (see https://concourse.halfpipe.io/teams/oscar/pipelines/oscar-sites-nature/jobs/Build/builds/1424). The build passed after that _old_ commit had been processed.

The next phase of the build pipeline didn't trigger for that commit. I am copying the out log from the git phase into a thread here.

```
git
ref    c3caf04e50d567058fa849ddf03bd5b8922e0dee
author    
Maria Rubio
author_date    
2019-05-31 13:20:54 +0100
commit    
c3caf04e50d567058fa849ddf03bd5b8922e0dee
committer    
Maria Rubio
committer_date    
2019-05-31 13:20:54 +0100
message    
[Maria] Prevent MailClient tests from running in a live environment

Identity added: /tmp/git-resource-private-key (/tmp/git-resource-private-key)
Cloning into '/tmp/build/get'...
Checking out files: 100% (9569/9569), done.
Could not find ref c3caf04e50d567058fa849ddf03bd5b8922e0dee in a shallow clone of depth 1
Deepening the shallow clone to depth 2...
Could not find ref c3caf04e50d567058fa849ddf03bd5b8922e0dee in a shallow clone of depth 2
Deepening the shallow clone to depth 4...
Could not find ref c3caf04e50d567058fa849ddf03bd5b8922e0dee in a shallow clone of depth 4
Deepening the shallow clone to depth 8...
Could not find ref c3caf04e50d567058fa849ddf03bd5b8922e0dee in a shallow clone of depth 8
Deepening the shallow clone to depth 16...
Could not find ref c3caf04e50d567058fa849ddf03bd5b8922e0dee in a shallow clone of depth 16
Deepening the shallow clone to depth 32...
Could not find ref c3caf04e50d567058fa849ddf03bd5b8922e0dee in a shallow clone of depth 32
Deepening the shallow clone to depth 64...
Could not find ref c3caf04e50d567058fa849ddf03bd5b8922e0dee in a shallow clone of depth 64
Deepening the shallow clone to depth 128...
Could not find ref c3caf04e50d567058fa849ddf03bd5b8922e0dee in a shallow clone of depth 128
Reached depth threshold 128, falling back to deep clone...
c3caf04e50 [Maria] Prevent MailClient tests from running in a live environment
```

**Comment by robwhitby** [comment]
So far we are at a loss to explain how this happened, other than on purpose - by pinning the version (or disabling all other versions) and manually triggering the build step. 

The update step does disable and enable versions depending on if the pipeline definition has changed. But even if this went wrong it wouldn't trigger a job with versions that have already passed.

**Comment by robwhitby** [comment]
Got report of this happening again from Maria Rubio

https://springernature.slack.com/archives/C9LFM8B6X/p1561986138197500

https://concourse.halfpipe.io/teams/oscar/pipelines/oscar-sites-bmc/jobs/Build/builds/1329

build ran at version `0.337.0`, build before was `0.434.0`. 

version `0.337.0` was successful when it first ran https://concourse.halfpipe.io/teams/oscar/pipelines/oscar-sites-bmc/jobs/Build/builds/1223

**Comment by mariarubiogonzalez** [comment]
One more example:
This commit triggered https://github.com/springernature/oscar/commit/fb2c771b931943db8e583d05c691b19365e8e433#diff-d87432d06a1a47967633bcc1ec292a27 a build of a previous commit happening 45 minutes before https://concourse.halfpipe.io/teams/oscar/pipelines/oscar-semi-pro/jobs/Build/builds/419

**Comment by robwhitby** [comment]
By querying the build history in the database I think I have managed to find all the jobs where this problem has happened. It is always the first job after the `update` job, which bumps the version resource.  Sometimes the next job has been triggered twice - first with the previous version (as if manually triggered), and then immediately after with the new version.

https://gist.github.com/robwhitby/1a01f27038c0817caafc368579ce66cd

Will continue investigating ...

**Comment by robwhitby** [comment]
progress update:

still not sure exactly what causes the bug, but have released a speculative change in halfpipe `3.9.0` and will monitor the effect. I have raised the issue with Concourse and also they are finishing off a new scheduling algorithm for release in the next version, so we will test that as soon as possible.
